aiobotocore/utils.py

```diff
@@ -367,30 +367,16 @@ async def get_credentials(self, **kwargs):
 
 
 class AioS3ExpressIdentityCache(AioIdentityCache, S3ExpressIdentityCache):
-    @functools.cached_property
-    def _aio_credential_cache(self):
-        """Substitutes upstream credential cache."""
-        return {}
+    @functools.lru_cache(maxsize=100)
+    def _get_credentials(self, bucket):
+        return asyncio.create_task(super().get_credentials(bucket=bucket))
 
     async def get_credentials(self, bucket):
         # upstream uses `@functools.lru_cache(maxsize=100)` to cache credentials.
         # This is incompatible with async code.
         # We need to implement custom caching logic.
 
-        if (credentials := self._aio_credential_cache.get(bucket)) is None:
-            # cache miss -> get credentials asynchronously
-            credentials = await super().get_credentials(bucket=bucket)
-
-            # upstream cache is bounded at 100 entries
-            if len(self._aio_credential_cache) >= 100:
-                # drop oldest entry from cache (deviates from lru_cache logic)
-                self._aio_credential_cache.pop(
-                    next(iter(self._aio_credential_cache)),
-                )
-
-            self._aio_credential_cache[bucket] = credentials
-
-        return credentials
+        return await self._get_credentials(bucket=bucket)
```

Comment:
I don't think this will work, you can have multiple calls that get the same task returned and all trying to await them.

Reply:
I beg to differ: calls to `self._get_credentials(bucket=bucket)` do not yield/await anything. The first call creates a task and caches it, while each subsequent call merely fetches the existing task from cache. The await always occurs after that. One could rewrite the statement as:

```Python
task = self._get_credentials(bucket=bucket)  # no await -> no concurrency possible
return await task  # await allows the event loop to run other tasks -> allows for concurrency, but race condition is mitigated
```

Reply:
Unless I'm missing something: Lets say you have 10 calls to `get_credentials` happening at the same time (GC1-GC10).  The first one to run will put a task (T1) into the LRU.  GC1 through GC10 will then get the same task returned (T1) and all try to await the same task (T1).  A task can only get await'ed once.

Reply:
I don't think this can work w/o synchronization

Reply:
> Unless I'm missing something: Lets say you have 10 calls to `get_credentials` happening at the same time (GC1-GC10). The first one to run will put a task (T1) into the LRU. GC1 through GC10 will then get the same task returned (T1) and all try to await the same task (T1). A task can only get await'ed once.

Sure, you can await a Task as often as you like. It will always return the result of the wrapped coroutine. It's like a mini-cache ;-) The wrapped coroutine is executed only once. No explicit synchronisation required.

Also, async code is not executed in parallel ("at the same time"), but concurrently (i.e. sequentially). For that reason, (explicit) synchronisation is not required as frequently as in multi-threaded code.

Reply:
aha, that was the crucial point, I did know you could await a task multiple times, for some reason I didn't think you could as logically you would think you can only execute the task once.  thanks!  That's a great benefit of that behavior as you used, built in synchronization! cool, I'll remove the lru part from my PR and just keep the venv enhancements